server/backends/migration_cache/migration_cache.go

```diff
@@ -121,46 +129,106 @@ func pebbleCacheFromConfig(env environment.Env, cfg *PebbleCacheConfig) (*pebble
-	return nil, status.UnimplementedError("not yet implemented")
-}
-
-func (mc MigrationCache) Set(ctx context.Context, d *repb.Digest, data []byte) error {
+func (mc *MigrationCache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error {
 	return status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error {
+func (mc *MigrationCache) Delete(ctx context.Context, d *repb.Digest) error {
 	return status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Delete(ctx context.Context, d *repb.Digest) error {
-	return status.UnimplementedError("not yet implemented")
+func (mc *MigrationCache) Reader(ctx context.Context, d *repb.Digest, offset, limit int64) (io.ReadCloser, error) {
+	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Reader(ctx context.Context, d *repb.Digest, offset, limit int64) (io.ReadCloser, error) {
+func (mc *MigrationCache) Writer(ctx context.Context, d *repb.Digest) (io.WriteCloser, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Writer(ctx context.Context, d *repb.Digest) (io.WriteCloser, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+// TODO(Maggie): Add copying logic
+func (mc *MigrationCache) Get(ctx context.Context, d *repb.Digest) ([]byte, error) {
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+	var srcBuf []byte
+
+	eg.Go(func() error {
+		srcBuf, srcErr = mc.src.Get(gctx, d)
+		return srcErr
+	})
+
+	// Double read some proportion to guarantee that data is consistent between caches
+	doubleRead := rand.Float64() <= mc.doubleReadPercentage
+	if doubleRead {
+		eg.Go(func() error {
+			_, dstErr = mc.dest.Get(gctx, d)
+			return nil // we don't care about the return error from this cache
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	if doubleRead {
+		if srcErr != dstErr {
+			// Don't log if data not found in dest cache, bc it may not have been copied yet
+			if !status.IsNotFoundError(dstErr) {
+				log.Warningf("Double read of %q failed. src err %s, dest err %s", d, srcErr, dstErr)
+			}
+		}
+	}
+
+	// Return data from source cache
+	return srcBuf, srcErr
+}
+
+func (mc *MigrationCache) Set(ctx context.Context, d *repb.Digest, data []byte) error {
+	eg, gctx := errgroup.WithContext(ctx)
+	var srcErr, dstErr error
+
+	// Double write data to both caches
+	eg.Go(func() error {
+		srcErr = mc.src.Set(gctx, d, data)
+		return srcErr
+	})
+
+	eg.Go(func() error {
+		dstErr = mc.dest.Set(gctx, d, data)
+		return nil // don't fail if there's an error from this cache
+	})
+
+	if err := eg.Wait(); err != nil {
+		// If error during write to source cache (source of truth), must delete from destination cache
+		deleteErr := mc.dest.Delete(ctx, d)
```

Comment:
do we only need to do this if there was not also an error from the destCache?

Reply:
Yes thanks - updated